interp/src/primitives/combinational.rs

```diff
@@ -67,37 +74,187 @@ comb_primitive!(StdNot[WIDTH](r#in: WIDTH) -> (out: WIDTH) {
 
 // ===================== Unsigned binary operations ======================
 comb_primitive!(StdAdd[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
-  let left_64 = left.as_u64();
-  let right_64 = right.as_u64();
-  let init_val = left_64 + right_64;
-  let bitwidth: usize = left.vec.len();
-  Value::from_init(init_val, bitwidth).into()
+    let a_iter = left.vec.iter().by_ref();
+    let b_iter = right.vec.iter().by_ref();
+    let mut c_in = false;
+    let mut sum = BitVec::new();
+    for (ai, bi) in a_iter.zip(b_iter) {
+        sum.push(
+            c_in & !ai & !bi
+                || bi & !c_in & !ai
+                || ai & !c_in & !bi
+                || ai & bi & c_in,
+        );
+        c_in = bi & c_in || ai & c_in || ai & bi || ai & c_in & bi;
+    }
+    let tr = Value { vec: sum };
+    //as a sanity check, check tr has same width as left
+    assert_eq!(tr.width(), left.width());
+    tr.into()
 });
 comb_primitive!(StdSub[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
-  let left_64 = left.as_u64();
-  // Bitwise subtraction: left - right = left + (!right + 1)
-  let right = Value {
-      vec: !(right.vec.clone()),
-  };
-  let right_64 = right.as_u64() + 1;
-  //need to allow overflow -- we are dealing only with bits
-  let init_val = left_64 + right_64;
-  let bitwidth: usize = left.vec.len();
-  Value::from_init(init_val, bitwidth).into()
+    //first turn right into ~right + 1
+    let new_right = !right.vec.clone();
+    let mut adder = StdAdd::from_constants(WIDTH + 1);
+    let new_right = adder
+        .execute(
+            &[("left".into(), &Value { vec: new_right }),
+            ("right".into(), &Value::from(1, WIDTH).unwrap())], None
+        )
+        .into_iter()
+        .next()
+        .map(|(_, v)| v)
+        .unwrap()
+        .unwrap_imm();
+
+    //now do addition. maybe better to use the adder and unwrap the OutputValue?
```

Comment:
Seems worth a shot, if it's not too hard! (Just to cut down on duplication a bit.)